File: p344e/reverse_string.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm -f test
#include <string>
#include <utility>
#include <vector>
#include <iostream>

using namespace std;
class Solution {
public:
    string reverseString(string s) {
        string ans = s;
        int l = 0, r = s.size() - 1;
        while (l < r) {
            swap(ans[l++], ans[r--]);
        }
        return ans;
    }
};

int main(int argc, char const *argv[])
{
    vector<string> tests = {
        "",
        "hello",
        "A man, a plan, a canal: Panama"
    };
    Solution sol;
    for (auto t: tests) {
        cout << "Reverse(\"" << t << "\") -> "
             << sol.reverseString(t) << endl;
        assert(t == sol.reverseString(sol.reverseString(t)));
    }
    return 0;
}

File: p344e/reverse_string.py
from typing import List


class Solution:
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        l, r = 0, len(s) - 1
        while l < r:
            s[l], s[r] = s[r], s[l]
            l, r = l + 1, r - 1


# TESTS
tests = [
    (["h", "e", "l", "l", "o"], ["o", "l", "l", "e", "h"]),
    (["H", "a", "n", "n", "a", "h"], ["h", "a", "n", "n", "a", "H"]),
]
for t in tests:
    sol = Solution()
    sol.reverseString(t[0])
    assert t[0] == t[1]
